File: Projects/Defensive/TLS setup/tls_toolkit.c
/**
 * TLS Setup & Hardening Guide - C Implementation
 * Compile: gcc -o tls_toolkit tls_toolkit.c -lssl -lcrypto
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <openssl/ssl.h>
#include <openssl/err.h>
#include <openssl/x509.h>
#include <openssl/x509v3.h>
#include <openssl/bio.h>
#include <time.h>

#define MAX_BUFFER 4096

typedef struct {
    char* name;
    char* protocols;
    char* ciphers;
    char* curves;
    int security_level;
} TLSConfig;

typedef struct {
    char* hostname;
    int port;
    char* tls_version;
    char* cipher;
    char* certificate_info;
    int vulnerabilities;
} TLSTestResult;

void initialize_openssl() {
    SSL_library_init();
    SSL_load_error_strings();
    OpenSSL_add_all_algorithms();
}

void cleanup_openssl() {
    EVP_cleanup();
}

TLSConfig* create_tls_config(const char* level) {
    TLSConfig* config = malloc(sizeof(TLSConfig));
    
    if(strcmp(level, "modern") == 0) {
        config->name = "Modern";
        config->protocols = "TLSv1.3 TLSv1.2";
        config->ciphers = "TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256";
        config->curves = "X25519:secp521r1:secp384r1";
        config->security_level = 3;
    } else if(strcmp(level, "intermediate") == 0) {
        config->name = "Intermediate";
        config->protocols = "TLSv1.2 TLSv1.3";
        config->ciphers = "ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305";
        config->curves = "prime256v1:secp384r1:secp521r1";
        config->security_level = 2;
    } else {
        config->name = "Compatible";
        config->protocols = "TLSv1.2 TLSv1.3";
        config->ciphers = "ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256";
        config->curves = "prime256v1:secp384r1:secp521r1";
        config->security_level = 1;
    }
    
    return config;
}

void generate_nginx_config(TLSConfig* config, const char* output_file) {
    FILE* fp = output_file ? fopen(output_file, "w") : stdout;
    
    if(!fp) {
        printf("Error: Cannot open output file\n");
        return;
    }
    
    fprintf(fp, "# TLS Hardening Configuration for Nginx\n");
    fprintf(fp, "# Security Level: %s\n", config->name);
    fprintf(fp, "# Generated: %s\n", __DATE__);
    fprintf(fp, "\n");
    
    fprintf(fp, "ssl_protocols %s;\n", config->protocols);
    fprintf(fp, "ssl_ciphers %s;\n", config->ciphers);
    fprintf(fp, "ssl_prefer_server_ciphers on;\n");
    fprintf(fp, "\n");
    
    fprintf(fp, "# Modern SSL settings\n");
    fprintf(fp, "ssl_session_timeout 1d;\n");
    fprintf(fp, "ssl_session_cache shared:SSL:50m;\n");
    fprintf(fp, "ssl_session_tickets off;\n");
    fprintf(fp, "\n");
    
    fprintf(fp, "# Security headers\n");
    fprintf(fp, "add_header Strict-Transport-Security \"max-age=63072000\" always;\n");
    fprintf(fp, "add_header X-Frame-Options DENY;\n");
    fprintf(fp, "add_header X-Content-Type-Options nosniff;\n");
    fprintf(fp, "add_header X-XSS-Protection \"1; mode=block\";\n");
    fprintf(fp, "\n");
    
    fprintf(fp, "# DH parameters\n");
    fprintf(fp, "ssl_dhparam /etc/nginx/dhparam.pem;\n");
    fprintf(fp, "\n");
    
    fprintf(fp, "# OCSP Stapling\n");
    fprintf(fp, "ssl_stapling on;\n");
    fprintf(fp, "ssl_stapling_verify on;\n");
    fprintf(fp, "resolver 8.8.8.8 8.8.4.4 valid=300s;\n");
    fprintf(fp, "resolver_timeout 5s;\n");
    fprintf(fp, "\n");
    
    fprintf(fp, "# ECDH curves\n");
    fprintf(fp, "ssl_ecdh_curve %s;\n", config->curves);
    
    if(output_file) {
        fclose(fp);
        printf("[+] Nginx configuration saved to: %s\n", output_file);
    }
}

void generate_apache_config(TLSConfig* config, const char* output_file) {
    FILE* fp = output_file ? fopen(output_file, "w") : stdout;
    
    if(!fp) {
        printf("Error: Cannot open output file\n");
        return;
    }
    
    fprintf(fp, "# TLS Hardening Configuration for Apache\n");
    fprintf(fp, "# Security Level: %s\n", config->name);
    fprintf(fp, "# Generated: %s\n", __DATE__);
    fprintf(fp, "\n");
    
    fprintf(fp, "SSLProtocol %s\n", config->protocols);
    fprintf(fp, "SSLCipherSuite %s\n", config->ciphers);
    fprintf(fp, "SSLHonorCipherOrder on\n");
    fprintf(fp, "\n");
    
    fprintf(fp, "# Modern SSL settings\n");
    fprintf(fp, "SSLSessionCache \"shmcb:/var/run/ssl_scache(512000)\"\n");
    fprintf(fp, "SSLSessionCacheTimeout 300\n");
    fprintf(fp, "\n");
    
    fprintf(fp, "# Security headers\n");
    fprintf(fp, "Header always set Strict-Transport-Security \"max-age=63072000\"\n");
    fprintf(fp, "Header always set X-Frame-Options DENY\n");
    fprintf(fp, "Header always set X-Content-Type-Options nosniff\n");
    fprintf(fp, "Header always set X-XSS-Protection \"1; mode=block\"\n");
    fprintf(fp, "\n");
    
    fprintf(fp, "# OCSP Stapling\n");
    fprintf(fp, "SSLUseStapling On\n");
    fprintf(fp, "SSLStaplingResponderTimeout 5\n");
    fprintf(fp, "SSLStaplingReturnResponderErrors Off\n");
    fprintf(fp, "SSLStaplingCache \"shmcb:/var/run/ocsp(128000)\"\n");
    
    if(output_file) {
        fclose(fp);
        printf("[+] Apache configuration saved to: %s\n", output_file);
    }
}

void test_tls_connection(const char* hostname, int port) {
    printf("[*] Testing TLS connection to %s:%d\n", hostname, port);
    
    SSL_CTX* ctx = SSL_CTX_new(TLS_client_method());
    if(!ctx) {
        printf("Error: Cannot create SSL context\n");
        return;
    }
    
    // Set minimum TLS version
    SSL_CTX_set_min_proto_version(ctx, TLS1_2_VERSION);
    
    BIO* bio = BIO_new_ssl_connect(ctx);
    if(!bio) {
        printf("Error: Cannot create BIO\n");
        SSL_CTX_free(ctx);
        return;
    }
    
    SSL* ssl;
    BIO_get_ssl(bio, &ssl);
    SSL_set_mode(ssl, SSL_MODE_AUTO_RETRY);
    
    // Set SNI
    SSL_set_tlsext_host_name(ssl, hostname);
    
    char hostport[256];
    snprintf(hostport, sizeof(hostport), "%s:%d", hostname, port);
    BIO_set_conn_hostname(bio, hostport);
    
    if(BIO_do_connect(bio) <= 0) {
        printf("Error: Cannot connect to %s\n", hostport);
        BIO_free_all(bio);
        SSL_CTX_free(ctx);
        return;
    }
    
    if(BIO_do_handshake(bio) <= 0) {
        printf("Error: TLS handshake failed\n");
        BIO_free_all(bio);
        SSL_CTX_free(ctx);
        return;
    }
    
    printf("[+] TLS connection established\n");
    
    // Get connection information
    printf("TLS Version: %s\n", SSL_get_version(ssl));
    
    const SSL_CIPHER* cipher = SSL_get_current_cipher(ssl);
    printf("Cipher Suite: %s\n", SSL_CIPHER_get_name(cipher));
    
    X509* cert = SSL_get_peer_certificate(ssl);
    if(cert) {
        printf("Certificate Subject: ");
        X509_NAME_print_ex_fp(stdout, X509_get_subject_name(cert), 0, XN_FLAG_ONELINE);
        printf("\n");
        
        printf("Certificate Issuer: ");
        X509_NAME_print_ex_fp(stdout, X509_get_issuer_name(cert), 0, XN_FLAG_ONELINE);
        printf("\n");
        
        // Check certificate expiration
        ASN1_TIME* not_after = X509_get_notAfter(cert);
        printf("Certificate Expires: ");
        ASN1_TIME_print_fp(stdout, not_after);
        printf("\n");
        
        X509_free(cert);
    }
    
    BIO_free_all(bio);
    SSL_CTX_free(ctx);
}

void print_usage() {
    printf("TLS Setup & Hardening Toolkit\n");
    printf("Usage:\n");
    printf("  Generate Nginx config:  tls_toolkit nginx <level> [output_file]\n");
    printf("  Generate Apache config: tls_toolkit apache <level> [output_file]\n");
    printf("  Test TLS connection:    tls_toolkit test <hostname> [port]\n");
    printf("\n");
    printf("Security levels: modern, intermediate, compatible\n");
    printf("Default port: 443\n");
}

int main(int argc, char* argv[]) {
    if(argc < 2) {
        print_usage();
        return 1;
    }
    
    initialize_openssl();
    
    if(strcmp(argv[1], "nginx") == 0 && argc >= 3) {
        TLSConfig* config = create_tls_config(argv[2]);
        generate_nginx_config(config, argc > 3 ? argv[3] : NULL);
        free(config);
    }
    else if(strcmp(argv[1], "apache") == 0 && argc >= 3) {
        TLSConfig* config = create_tls_config(argv[2]);
        generate_apache_config(config, argc > 3 ? argv[3] : NULL);
        free(config);
    }
    else if(strcmp(argv[1], "test") == 0 && argc >= 3) {
        int port = 443;
        if(argc > 3) {
            port = atoi(argv[3]);
        }
        test_tls_connection(argv[2], port);
    }
    else {
        print_usage();
    }
    
    cleanup_openssl();
    return 0;
}

File: Projects/Defensive/TLS setup/tls_toolkit.py
#!/usr/bin/env python3
"""
TLS Setup & Hardening Guide - Python Implementation
Secure TLS configuration generator and tester
"""

import ssl
import socket
import OpenSSL
import json
import yaml
import argparse
import subprocess
import sys
from datetime import datetime
from pathlib import Path
from typing import Dict, List, Any, Optional
import hashlib

class TLSConfigGenerator:
    """Generate secure TLS configurations for various servers"""
    
    def __init__(self):
        self.cipher_suites = {
            'modern': [
                'TLS_AES_256_GCM_SHA384',
                'TLS_CHACHA20_POLY1305_SHA256', 
                'TLS_AES_128_GCM_SHA256',
                'TLS_AES_128_CCM_SHA256'
            ],
            'intermediate': [
                'ECDHE-ECDSA-AES256-GCM-SHA384',
                'ECDHE-RSA-AES256-GCM-SHA384',
                'ECDHE-ECDSA-CHACHA20-POLY1305',
                'ECDHE-RSA-CHACHA20-POLY1305',
                'ECDHE-ECDSA-AES128-GCM-SHA256',
                'ECDHE-RSA-AES128-GCM-SHA256',
                'DHE-RSA-AES256-GCM-SHA384',
                'DHE-RSA-AES128-GCM-SHA256'
            ],
            'compatible': [
                'ECDHE-ECDSA-AES256-GCM-SHA384',
                'ECDHE-RSA-AES256-GCM-SHA384',
                'ECDHE-ECDSA-CHACHA20-POLY1305',
                'ECDHE-RSA-CHACHA20-POLY1305',
                'ECDHE-ECDSA-AES128-GCM-SHA256',
                'ECDHE-RSA-AES128-GCM-SHA256',
                'DHE-RSA-AES256-GCM-SHA384',
                'DHE-RSA-AES128-GCM-SHA256',
                'ECDHE-ECDSA-AES256-SHA384',
                'ECDHE-RSA-AES256-SHA384',
                'ECDHE-ECDSA-AES128-SHA256',
                'ECDHE-RSA-AES128-SHA256'
            ]
        }
        
        self.tls_versions = {
            'modern': ['TLSv1.3', 'TLSv1.2'],
            'intermediate': ['TLSv1.2', 'TLSv1.3'],
            'compatible': ['TLSv1.2', 'TLSv1.3', 'TLSv1.1']
        }
        
        self.curves = {
            'modern': ['X25519', 'secp521r1', 'secp384r1'],
            'intermediate': ['prime256v1', 'secp384r1', 'secp521r1'],
            'compatible': ['prime256v1', 'secp384r1', 'secp521r1']
        }
    
    def generate_nginx_config(self, security_level: str = 'intermediate') -> str:
        """Generate secure Nginx TLS configuration"""
        config = f"""
# TLS Hardening Configuration for Nginx
# Security Level: {security_level}
# Generated: {datetime.now().isoformat()}

ssl_protocols {' '.join(self.tls_versions[security_level])};
ssl_ciphers {' '.join(self.cipher_suites[security_level])};
ssl_prefer_server_ciphers on;

# Modern SSL settings
ssl_session_timeout 1d;
ssl_session_cache shared:SSL:50m;
ssl_session_tickets off;

# Modern security headers
add_header Strict-Transport-Security "max-age=63072000" always;
add_header X-Frame-Options DENY;
add_header X-Content-Type-Options nosniff;
add_header X-XSS-Protection "1; mode=block";

# DH parameters (generate with: openssl dhparam -out dhparam.pem 4096)
ssl_dhparam /etc/nginx/dhparam.pem;

# OCSP Stapling
ssl_stapling on;
ssl_stapling_verify on;
resolver 8.8.8.8 8.8.4.4 valid=300s;
resolver_timeout 5s;

# Security enhancements
ssl_ecdh_curve {' '.join(self.curves[security_level])};
"""
        return config
    
    def generate_apache_config(self, security_level: str = 'intermediate') -> str:
        """Generate secure Apache TLS configuration"""
        config = f"""
# TLS Hardening Configuration for Apache
# Security Level: {security_level}
# Generated: {datetime.now().isoformat()}

SSLProtocol {' '.join(self.tls_versions[security_level])}
SSLCipherSuite {' '.join(self.cipher_suites[security_level])}
SSLHonorCipherOrder on

# Modern SSL settings
SSLSessionCache "shmcb:/var/run/ssl_scache(512000)"
SSLSessionCacheTimeout 300

# Security headers
Header always set Strict-Transport-Security "max-age=63072000"
Header always set X-Frame-Options DENY
Header always set X-Content-Type-Options nosniff
Header always set X-XSS-Protection "1; mode=block"

# OCSP Stapling
SSLUseStapling On
SSLStaplingResponderTimeout 5
SSLStaplingReturnResponderErrors Off
SSLStaplingCache "shmcb:/var/run/ocsp(128000)"
"""
        return config
    
    def generate_openssl_config(self) -> str:
        """Generate secure OpenSSL configuration"""
        config = """
# Secure OpenSSL Configuration
[openssl_init]
ssl_conf = ssl_section

[ssl_section]
system_default = system_default_section

[system_default_section]
MinProtocol = TLSv1.2
CipherString = DEFAULT@SECLEVEL=2
Options = UnsafeLegacyRenegotiation
"""
        return config
    
    def generate_python_ssl_context(self, security_level: str = 'intermediate') -> str:
        """Generate Python SSL context configuration"""
        config = f'''
import ssl

def create_secure_ssl_context():
    """Create a secure SSL context"""
    context = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH)
    
    # TLS versions
    context.minimum_version = ssl.TLSVersion.{self.tls_versions[security_level][-1].replace("TLSv", "TLS").replace(".", "")}
    context.maximum_version = ssl.TLSVersion.{self.tls_versions[security_level][0].replace("TLSv", "TLS").replace(".", "")}
    
    # Cipher suites
    if hasattr(context, 'set_ciphers'):
        context.set_ciphers(':'.join({self.cipher_suites[security_level]}))
    
    # Security options
    context.options |= ssl.OP_NO_SSLv2
    context.options |= ssl.OP_NO_SSLv3
    context.options |= ssl.OP_NO_TLSv1
    context.options |= ssl.OP_NO_TLSv1_1
    context.options |= ssl.OP_NO_COMPRESSION
    context.options |= ssl.OP_SINGLE_DH_USE
    context.options |= ssl.OP_SINGLE_ECDH_USE
    
    # Certificate verification
    context.check_hostname = True
    context.verify_mode = ssl.CERT_REQUIRED
    
    return context
'''
        return config

class TLSCertificateManager:
    """Manage TLS certificates and keys"""
    
    def __init__(self, cert_dir: str = "./certs"):
        self.cert_dir = Path(cert_dir)
        self.cert_dir.mkdir(exist_ok=True)
    
    def generate_self_signed_cert(self, common_name: str, days: int = 365) -> Dict[str, str]:
        """Generate self-signed certificate for testing"""
        try:
            from OpenSSL import crypto
            
            # Generate key
            key = crypto.PKey()
            key.generate_key(crypto.TYPE_RSA, 4096)
            
            # Generate certificate
            cert = crypto.X509()
            cert.get_subject().CN = common_name
            cert.set_serial_number(1000)
            cert.gmtime_adj_notBefore(0)
            cert.gmtime_adj_notAfter(days * 24 * 60 * 60)
            cert.set_issuer(cert.get_subject())
            cert.set_pubkey(key)
            cert.sign(key, 'sha256')
            
            # Save files
            key_path = self.cert_dir / "key.pem"
            cert_path = self.cert_dir / "cert.pem"
            
            with open(key_path, "wb") as f:
                f.write(crypto.dump_privatekey(crypto.FILETYPE_PEM, key))
            
            with open(cert_path, "wb") as f:
                f.write(crypto.dump_certificate(crypto.FILETYPE_PEM, cert))
            
            return {
                'key_file': str(key_path),
                'cert_file': str(cert_path),
                'common_name': common_name,
                'expires': days
            }
            
        except ImportError:
            print("PyOpenSSL not available. Using fallback method.")
            return self._generate_cert_fallback(common_name, days)
    
    def _generate_cert_fallback(self, common_name: str, days: int) -> Dict[str, str]:
        """Fallback certificate generation using openssl command"""
        key_path = self.cert_dir / "key.pem"
        cert_path = self.cert_dir / "cert.pem"
        
        # Generate private key
        subprocess.run([
            'openssl', 'genrsa', '-out', str(key_path), '4096'
        ], check=True)
        
        # Generate self-signed certificate
        subprocess.run([
            'openssl', 'req', '-new', '-x509', '-key', str(key_path),
            '-out', str(cert_path), '-days', str(days),
            '-subj', f'/CN={common_name}'
        ], check=True)
        
        return {
            'key_file': str(key_path),
            'cert_file': str(cert_path),
            'common_name': common_name,
            'expires': days
        }
    
    def analyze_certificate(self, cert_file: str) -> Dict[str, Any]:
        """Analyze TLS certificate for security issues"""
        try:
            with open(cert_file, 'rb') as f:
                cert_data = f.read()
            
            cert = OpenSSL.crypto.load_certificate(OpenSSL.crypto.FILETYPE_PEM, cert_data)
            
            analysis = {
                'subject': dict(cert.get_subject().get_components()),
                'issuer': dict(cert.get_issuer().get_components()),
                'serial_number': cert.get_serial_number(),
                'version': cert.get_version(),
                'not_before': cert.get_notBefore().decode('utf-8'),
                'not_after': cert.get_notAfter().decode('utf-8'),
                'signature_algorithm': cert.get_signature_algorithm().decode('utf-8'),
                'key_bits': cert.get_pubkey().bits(),
                'key_type': 'RSA' if cert.get_pubkey().type() == OpenSSL.crypto.TYPE_RSA else 'Unknown',
                'extensions': []
            }
            
            # Check extensions
            for i in range(cert.get_extension_count()):
                ext = cert.get_extension(i)
                analysis['extensions'].append({
                    'name': ext.get_short_name().decode('utf-8'),
                    'critical': ext.get_critical(),
                    'data': str(ext)
                })
            
            # Security checks
            analysis['security_checks'] = self._perform_certificate_checks(cert)
            
            return analysis
            
        except Exception as e:
            return {'error': str(e)}
    
    def _perform_certificate_checks(self, cert) -> Dict[str, Any]:
        """Perform security checks on certificate"""
        checks = {}
        
        # Check key size
        key_bits = cert.get_pubkey().bits()
        checks['key_size_adequate'] = key_bits >= 2048
        
        # Check validity period
        not_after = datetime.strptime(cert.get_notAfter().decode('utf-8'), '%Y%m%d%H%M%SZ')
        validity_days = (not_after - datetime.now()).days
        checks['validity_period_reasonable'] = validity_days <= 825  # ~2 years
        
        # Check signature algorithm
        sig_algo = cert.get_signature_algorithm().decode('utf-8')
        checks['strong_signature_algorithm'] = 'sha256' in sig_algo.lower() or 'sha384' in sig_algo.lower() or 'sha512' in sig_algo.lower()
        
        return checks

class TLSTester:
    """Test TLS configurations and security"""
    
    def __init__(self):
        self.test_results = {}
    
    def test_tls_server(self, hostname: str, port: int = 443) -> Dict[str, Any]:
        """Test TLS configuration of a remote server"""
        try:
            context = ssl.create_default_context()
            with socket.create_connection((hostname, port), timeout=10) as sock:
                with context.wrap_socket(sock, server_hostname=hostname) as ssock:
                    cert = ssock.getpeercert()
                    cipher = ssock.cipher()
                    version = ssock.version()
                    
                    result = {
                        'hostname': hostname,
                        'port': port,
                        'tls_version': version,
                        'cipher_suite': cipher,
                        'certificate': cert,
                        'supported_protocols': self._test_supported_protocols(hostname, port),
                        'security_issues': self._check_security_issues(version, cipher, cert)
                    }
                    
                    return result
                    
        except Exception as e:
            return {'error': str(e)}
    
    def _test_supported_protocols(self, hostname: str, port: int) -> Dict[str, bool]:
        """Test which TLS protocols are supported"""
        protocols = {
            'SSLv2': False,
            'SSLv3': False, 
            'TLSv1.0': False,
            'TLSv1.1': False,
            'TLSv1.2': False,
            'TLSv1.3': False
        }
        
        # Test each protocol
        for protocol_name in protocols.keys():
            try:
                context = ssl.SSLContext(ssl.PROTOCOL_SSLv23)  # Modern context
                
                # Set specific protocol version
                if protocol_name == 'SSLv2':
                    context.options |= ssl.OP_NO_SSLv3 | ssl.OP_NO_TLSv1 | ssl.OP_NO_TLSv1_1 | ssl.OP_NO_TLSv1_2
                elif protocol_name == 'SSLv3':
                    context.options |= ssl.OP_NO_SSLv2 | ssl.OP_NO_TLSv1 | ssl.OP_NO_TLSv1_1 | ssl.OP_NO_TLSv1_2
                elif protocol_name == 'TLSv1.0':
                    context.options |= ssl.OP_NO_SSLv2 | ssl.OP_NO_SSLv3 | ssl.OP_NO_TLSv1_1 | ssl.OP_NO_TLSv1_2
                elif protocol_name == 'TLSv1.1':
                    context.options |= ssl.OP_NO_SSLv2 | ssl.OP_NO_SSLv3 | ssl.OP_NO_TLSv1 | ssl.OP_NO_TLSv1_2
                elif protocol_name == 'TLSv1.2':
                    context.minimum_version = ssl.TLSVersion.TLSv1_2
                    context.maximum_version = ssl.TLSVersion.TLSv1_2
                elif protocol_name == 'TLSv1.3':
                    context.minimum_version = ssl.TLSVersion.TLSv1_3
                    context.maximum_version = ssl.TLSVersion.TLSv1_3
                
                with socket.create_connection((hostname, port), timeout=5) as sock:
                    with context.wrap_socket(sock, server_hostname=hostname) as ssock:
                        protocols[protocol_name] = True
                        
            except:
                protocols[protocol_name] = False
        
        return protocols
    
    def _check_security_issues(self, version: str, cipher: tuple, cert: dict) -> List[str]:
        """Check for common security issues"""
        issues = []
        
        # Check TLS version
        if version in ['SSLv2', 'SSLv3', 'TLSv1.0', 'TLSv1.1']:
            issues.append(f'Weak TLS version: {version}')
        
        # Check cipher suite
        cipher_name = cipher[0] if cipher else ''
        weak_ciphers = ['RC4', 'MD5', 'SHA1', 'NULL', 'EXPORT', 'ANON', 'ADH', '3DES']
        if any(weak in cipher_name for weak in weak_ciphers):
            issues.append(f'Weak cipher suite: {cipher_name}')
        
        # Check certificate
        if cert:
            # Check key size
            key_size = cert.get('keySize', 0)
            if key_size < 2048:
                issues.append(f'Weak RSA key size: {key_size} bits')
            
            # Check signature algorithm
            sig_algo = cert.get('signatureAlgorithm', '')
            if 'sha1' in sig_algo.lower():
                issues.append('Weak signature algorithm: SHA1')
        
        return issues
    
    def scan_tls_vulnerabilities(self, hostname: str, port: int = 443) -> Dict[str, Any]:
        """Scan for common TLS vulnerabilities"""
        vulnerabilities = {
            'poodle': self._test_poodle(hostname, port),
            'heartbleed': self._test_heartbleed(hostname, port),
            'freak': self._test_freak(hostname, port),
            'logjam': self._test_logjam(hostname, port),
            'drown': self._test_drown(hostname, port)
        }
        
        return vulnerabilities
    
    def _test_poodle(self, hostname: str, port: int) -> Dict[str, Any]:
        """Test for POODLE vulnerability"""
        try:
            context = ssl.SSLContext(ssl.PROTOCOL_SSLv3)
            with socket.create_connection((hostname, port), timeout=5) as sock:
                with context.wrap_socket(sock, server_hostname=hostname):
                    return {'vulnerable': True, 'description': 'SSLv3 enabled - vulnerable to POODLE'}
        except:
            return {'vulnerable': False, 'description': 'SSLv3 not supported'}
    
    def _test_heartbleed(self, hostname: str, port: int) -> Dict[str, Any]:
        """Test for Heartbleed vulnerability (simplified)"""
        # Note: This is a basic test. Use dedicated tools for comprehensive testing.
        try:
            context = ssl.create_default_context()
            with socket.create_connection((hostname, port), timeout=5) as sock:
                with context.wrap_socket(sock, server_hostname=hostname):
                    return {'vulnerable': False, 'description': 'No Heartbleed detected (basic test)'}
        except Exception as e:
            return {'vulnerable': 'Unknown', 'description': f'Test inconclusive: {e}'}
    
    def _test_freak(self, hostname: str, port: int) -> Dict[str, Any]:
        """Test for FREAK vulnerability"""
        return {'vulnerable': False, 'description': 'FREAK test requires specialized tools'}
    
    def _test_logjam(self, hostname: str, port: int) -> Dict[str, Any]:
        """Test for Logjam vulnerability"""
        return {'vulnerable': False, 'description': 'Logjam test requires specialized tools'}
    
    def _test_drown(self, hostname: str, port: int) -> Dict[str, Any]:
        """Test for DROWN vulnerability"""
        return {'vulnerable': False, 'description': 'DROWN test requires specialized tools'}

def main():
    parser = argparse.ArgumentParser(description='TLS Setup & Hardening Guide')
    subparsers = parser.add_subparsers(dest='command', help='Command to execute')
    
    # Generate config command
    config_parser = subparsers.add_parser('generate', help='Generate TLS configurations')
    config_parser.add_argument('--server', choices=['nginx', 'apache', 'openssl', 'python'], 
                              required=True, help='Server type')
    config_parser.add_argument('--level', choices=['modern', 'intermediate', 'compatible'],
                              default='intermediate', help='Security level')
    config_parser.add_argument('--output', help='Output file')
    
    # Certificate management command
    cert_parser = subparsers.add_parser('certificate', help='Certificate management')
    cert_parser.add_argument('--generate', action='store_true', help='Generate self-signed certificate')
    cert_parser.add_argument('--analyze', help='Analyze certificate file')
    cert_parser.add_argument('--common-name', default='localhost', help='Common name for certificate')
    cert_parser.add_argument('--days', type=int, default=365, help='Certificate validity days')
    
    # Testing command
    test_parser = subparsers.add_parser('test', help='Test TLS configuration')
    test_parser.add_argument('hostname', help='Hostname to test')
    test_parser.add_argument('--port', type=int, default=443, help='Port to test')
    test_parser.add_argument('--scan-vulnerabilities', action='store_true', 
                           help='Scan for common vulnerabilities')
    
    args = parser.parse_args()
    
    if args.command == 'generate':
        generator = TLSConfigGenerator()
        
        if args.server == 'nginx':
            config = generator.generate_nginx_config(args.level)
        elif args.server == 'apache':
            config = generator.generate_apache_config(args.level)
        elif args.server == 'openssl':
            config = generator.generate_openssl_config()
        elif args.server == 'python':
            config = generator.generate_python_ssl_context(args.level)
        
        if args.output:
            with open(args.output, 'w') as f:
                f.write(config)
            print(f"[+] Configuration saved to {args.output}")
        else:
            print(config)
    
    elif args.command == 'certificate':
        manager = TLSCertificateManager()
        
        if args.generate:
            result = manager.generate_self_signed_cert(args.common_name, args.days)
            print(f"[+] Generated certificate:")
            print(f"    Key file: {result['key_file']}")
            print(f"    Cert file: {result['cert_file']}")
            print(f"    Common Name: {result['common_name']}")
            print(f"    Expires in: {result['expires']} days")
        
        if args.analyze:
            analysis = manager.analyze_certificate(args.analyze)
            print(json.dumps(analysis, indent=2))
    
    elif args.command == 'test':
        tester = TLSTester()
        
        if args.scan_vulnerabilities:
            vulnerabilities = tester.scan_tls_vulnerabilities(args.hostname, args.port)
            print("TLS Vulnerability Scan Results:")
            print(json.dumps(vulnerabilities, indent=2))
        else:
            result = tester.test_tls_server(args.hostname, args.port)
            print("TLS Configuration Test Results:")
            print(json.dumps(result, indent=2))
    
    else:
        parser.print_help()

if __name__ == "__main__":
    main()
